File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-0.java
class Solution {

    private boolean[][] pacific;

    private boolean[][] atlantic;

    private int[][] heights;

    private int rows;

    private int cols;

    private final int[][] directions = {{1,0},{-1,0},{0,1},{0,-1}};

    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        
        this.heights = heights;

        rows = heights.length;

        cols = heights[0].length;

        pacific = new boolean[rows][cols];

        atlantic = new boolean[rows][cols];

        List<List<Integer>> result = new ArrayList<>();

        for(int r=0; r < rows; r++){

            helper(r,0,pacific);
        }
        for(int c=0; c < cols; c++){

            helper(0,c,pacific);
        }

        for(int r= 0; r < rows; r++){

            helper(r,cols-1,atlantic);
        }

        for(int c= 0; c < cols; c++){

            helper(rows-1,c,atlantic);
        }

        for(int r=0; r < rows; r++){

            for(int c=0; c < cols; c++){

                if(pacific[r][c] && atlantic[r][c]){

                    result.add(Arrays.asList(r,c));
                }
            }
        }
        return result;
    }

    private void helper(int r, int c, boolean[][] visited){

        if(visited[r][c]) return;

        visited[r][c] = true;

        for(int[] d : directions){

            int nr = r + d[0];

            int nc = c + d[1];

            if(nr < 0 || nr >= rows || nc < 0 || nc >= cols){

                continue;
            }

            if(heights[nr][nc] >= heights[r][c]){

                helper(nr,nc,visited);
            }
        }
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-0.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:

        if not grid:
            return 0

        rows = len(grid)
        cols = len(grid[0])

        max_area = 0

        def helper(r: int, c: int)->int:

            if r < 0 or r >= rows or c < 0 or c >= cols:
                return 0

            if grid[r][c] == 0:
                return 0

            grid[r][c] = 0

            return 1 + helper(r+1,c) + helper(r-1,c) + helper(r,c+1) + helper(r,c-1)

        
        for i in range(rows):
            for j in range(cols):

                if grid[i][j] == 1:
                    
                    max_area = max(max_area,helper(i,j))

        return max_area